File: use-case-financial-services/2-prepare/data_preparation_studio.py
"""
Step 2: Data Preparation (Python)
==================================
Load, profile, and join the four PremierBank datasets into an
Analytical Base Table (ABT) for loan default prediction.

Run this script in a Python session inside SAS Viya Workbench.

Outputs:
  - data/financial_services_abt.csv
  - Console: data cards, summary statistics, default distribution
"""

import pandas as pd
import numpy as np
import os

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------
DATA_DIR = SAS.symget('_USERHOME') + '/sas-hackathon-boot-camp-2026/use-case-financial-services/data'
OUTPUT_PATH = os.path.join(DATA_DIR, 'financial_services_abt.csv')

# ============================================================================
# 1. LOAD THE DATA
# ============================================================================
print("=" * 70)
print("STEP 2: DATA PREPARATION  (Python)")
print("=" * 70)

loan_applications = pd.read_csv(os.path.join(DATA_DIR, 'loan_applications.csv'))
credit_history    = pd.read_csv(os.path.join(DATA_DIR, 'credit_history.csv'))
employment        = pd.read_csv(os.path.join(DATA_DIR, 'employment.csv'))
payment_history   = pd.read_csv(os.path.join(DATA_DIR, 'payment_history.csv'))

print("\n--- Loaded Datasets ---")
print(f"  loan_applications: {loan_applications.shape[0]:>6,} rows x {loan_applications.shape[1]} cols")
print(f"  credit_history:    {credit_history.shape[0]:>6,} rows x {credit_history.shape[1]} cols")
print(f"  employment:        {employment.shape[0]:>6,} rows x {employment.shape[1]} cols")
print(f"  payment_history:   {payment_history.shape[0]:>6,} rows x {payment_history.shape[1]} cols")

# ============================================================================
# 2. DATA CARD — one per table
# ============================================================================
def print_data_card(name, df):
    """Print a concise data card for a DataFrame."""
    print(f"\n{'─' * 70}")
    print(f"DATA CARD: {name}")
    print(f"{'─' * 70}")
    print(f"  Rows: {len(df):,}    Columns: {len(df.columns)}")
    print(f"\n  {'Column':<30} {'Dtype':<12} {'Missing':>8}")
    print(f"  {'─'*30} {'─'*12} {'─'*8}")
    for col in df.columns:
        print(f"  {col:<30} {str(df[col].dtype):<12} {df[col].isna().sum():>8}")
    print(f"\n  Sample rows (first 3):")
    print(df.head(3).to_string(index=False, max_colwidth=20))

print("\n" + "=" * 70)
print("DATA CARDS")
print("=" * 70)

print_data_card("loan_applications.csv", loan_applications)
print_data_card("credit_history.csv", credit_history)
print_data_card("employment.csv", employment)
print_data_card("payment_history.csv", payment_history)

# ============================================================================
# 3. BASIC SUMMARY STATISTICS
# ============================================================================
print("\n" + "=" * 70)
print("SUMMARY STATISTICS")
print("=" * 70)

for name, df in [("loan_applications", loan_applications),
                 ("credit_history", credit_history),
                 ("employment", employment),
                 ("payment_history", payment_history)]:
    numeric_cols = df.select_dtypes(include='number').columns.tolist()
    if numeric_cols:
        print(f"\n--- {name} (numeric columns) ---")
        print(df[numeric_cols].describe().round(2).to_string())

    cat_cols = df.select_dtypes(include='object').columns.tolist()
    cat_cols = [c for c in cat_cols if not c.endswith('_id') and 'date' not in c.lower()]
    if cat_cols:
        print(f"\n--- {name} (categorical frequencies) ---")
        for col in cat_cols:
            print(f"\n  {col}:")
            print(df[col].value_counts().head(10).to_string())

# ============================================================================
# 4. CLEAN & PARSE DATES
# ============================================================================
print("\n" + "=" * 70)
print("DATA CLEANING")
print("=" * 70)

loan_applications['application_date'] = pd.to_datetime(
    loan_applications['application_date'], errors='coerce')
payment_history['payment_date'] = pd.to_datetime(
    payment_history['payment_date'], errors='coerce')

loan_applications['defaulted'] = loan_applications['defaulted'].astype(int)

# Validate loan_id consistency
valid_ids = set(loan_applications['loan_id'])
credit_history  = credit_history[credit_history['loan_id'].isin(valid_ids)]
employment      = employment[employment['loan_id'].isin(valid_ids)]
payment_history = payment_history[payment_history['loan_id'].isin(valid_ids)]

print("  Date columns parsed, IDs validated.")

# ============================================================================
# 5. FEATURE ENGINEERING
# ============================================================================
print("\n" + "=" * 70)
print("FEATURE ENGINEERING")
print("=" * 70)

# --- Payment behavior features ---------------------------------------------
pmt = payment_history.groupby('loan_id').agg(
    total_payments=('payment_id', 'count'),
    late_payment_count=('days_late', lambda x: (x > 0).sum()),
    severe_delinquency_count=('days_late', lambda x: (x >= 60).sum()),
    avg_days_late=('days_late', 'mean'),
    max_days_late=('days_late', 'max'),
    total_shortfall=('amount_due', lambda x: np.maximum(
        (x - payment_history.loc[x.index, 'amount_paid']).sum(), 0)),
).reset_index()

# Compute additional payment metrics
pmt_on_time = payment_history[payment_history['payment_status'] == 'On Time'].groupby(
    'loan_id').size().reset_index(name='on_time_count')
pmt = pmt.merge(pmt_on_time, on='loan_id', how='left')
pmt['on_time_count'] = pmt['on_time_count'].fillna(0).astype(int)

pmt['late_payment_rate'] = np.where(
    pmt['total_payments'] > 0,
    pmt['late_payment_count'] / pmt['total_payments'], 0)
pmt['on_time_rate'] = np.where(
    pmt['total_payments'] > 0,
    pmt['on_time_count'] / pmt['total_payments'], 0)
pmt['severe_delinquency_flag'] = (pmt['severe_delinquency_count'] > 0).astype(int)

# Average shortfall ratio per loan
shortfall = payment_history.copy()
shortfall['shortfall_ratio'] = np.where(
    shortfall['amount_due'] > 0,
    (shortfall['amount_due'] - shortfall['amount_paid']) / shortfall['amount_due'], 0)
avg_shortfall = shortfall.groupby('loan_id')['shortfall_ratio'].mean().reset_index(
    name='avg_shortfall_ratio')
pmt = pmt.merge(avg_shortfall, on='loan_id', how='left')
pmt['avg_shortfall_ratio'] = pmt['avg_shortfall_ratio'].fillna(0)

print(f"  Payment behavior features: {len(pmt.columns) - 1}")

# --- Credit features -------------------------------------------------------
credit = credit_history.copy()

# FICO score bands
credit['fico_band'] = pd.cut(
    credit['credit_score'],
    bins=[0, 599, 649, 699, 749, 850],
    labels=['Very Poor', 'Poor', 'Fair', 'Good', 'Excellent'])

# Credit utilization category
credit['utilization_cat'] = pd.cut(
    credit['credit_utilization'],
    bins=[-0.01, 0.30, 0.50, 0.75, 1.01],
    labels=['Low', 'Moderate', 'High', 'Very High'])

# Derived flags
credit['has_bankruptcy'] = (credit['bankruptcies'] > 0).astype(int)
credit['has_delinquency'] = (credit['delinquencies_2yr'] > 0).astype(int)
credit['high_inquiries'] = (credit['inquiries_6mo'] >= 3).astype(int)

print(f"  Credit features:           {len(credit.columns) - 1}")

# --- Employment features ---------------------------------------------------
emp = employment.copy()

# Debt service ratio
emp['debt_service_ratio'] = np.where(
    emp['annual_income'] > 0,
    (emp['monthly_debt'] * 12) / emp['annual_income'], 0)

# Income bands
emp['income_band'] = pd.cut(
    emp['annual_income'],
    bins=[0, 34999, 59999, 99999, float('inf')],
    labels=['Low', 'Low-Middle', 'Middle', 'High'])

# Years employed bands
emp['tenure_band'] = pd.cut(
    emp['years_employed'],
    bins=[-0.01, 1.99, 4.99, 9.99, float('inf')],
    labels=['New', 'Developing', 'Established', 'Veteran'])

print(f"  Employment features:       {len(emp.columns) - 1}")

# ============================================================================
# 6. BUILD THE ANALYTICAL BASE TABLE
# ============================================================================
print("\n" + "=" * 70)
print("BUILDING ANALYTICAL BASE TABLE")
print("=" * 70)

abt = loan_applications.copy()
abt.drop(columns=['application_date'], inplace=True)

abt = abt.merge(credit, on='loan_id', how='left')
abt = abt.merge(emp, on='loan_id', how='left')
abt = abt.merge(pmt, on='loan_id', how='left')

# Fill missing payment features
pmt_cols = [c for c in pmt.columns if c != 'loan_id']
for col in pmt_cols:
    abt[col] = abt[col].fillna(0)

# Fill missing debt_service_ratio
abt['debt_service_ratio'] = abt['debt_service_ratio'].fillna(0)

# One-hot encode categoricals
abt = pd.get_dummies(abt, columns=[
    'loan_purpose', 'property_type', 'employment_status', 'employer_type',
    'fico_band', 'utilization_cat', 'income_band', 'tenure_band'
], prefix=[
    'purpose', 'prop', 'emp', 'emplr',
    'fico', 'util', 'inc', 'ten'
])

# Convert boolean dummy columns to int
bool_cols = abt.select_dtypes(include='bool').columns
abt[bool_cols] = abt[bool_cols].astype(int)

print(f"  Final ABT shape: {abt.shape[0]:,} rows x {abt.shape[1]} columns")
print(f"  Features: {abt.shape[1] - 2}  (excluding loan_id and defaulted)")

# ============================================================================
# 7. DEFAULT DISTRIBUTION
# ============================================================================
print("\n" + "=" * 70)
print("DEFAULT DISTRIBUTION IN ABT")
print("=" * 70)

default_counts = abt['defaulted'].value_counts()
default_pct = abt['defaulted'].value_counts(normalize=True) * 100
print(f"  Current (0):   {default_counts[0]:>5,}  ({default_pct[0]:.1f}%)")
print(f"  Defaulted (1): {default_counts[1]:>5,}  ({default_pct[1]:.1f}%)")

# ============================================================================
# 8. SAVE
# ============================================================================
abt.to_csv(OUTPUT_PATH, index=False)
print(f"\nAnalytical Base Table saved to: {OUTPUT_PATH}")
print("=" * 70)
print("STEP 2 COMPLETE")
print("=" * 70)

File: use-case-financial-services/2-prepare/data_preparation_studio.R
# =============================================================================
# Step 2: Data Preparation (R)
# =============================================================================
# Load, profile, and join the four PremierBank datasets into an
# Analytical Base Table (ABT) for loan default prediction.
#
# Run this script in an R session inside SAS Studio.
# Uses only base R — no external packages required.
#
# Outputs:
#   - data/financial_services_abt.csv
#   - Console: data cards, summary statistics, default distribution
# =============================================================================

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------
data_dir    <- paste(symget("_USERHOME"), "sas-hackathon-boot-camp-2026/use-case-financial-services/data", sep="/")
output_path <- file.path(data_dir, "financial_services_abt.csv")

# ============================================================================
# 1. LOAD THE DATA
# ============================================================================
cat(strrep("=", 70), "\n")
cat("STEP 2: DATA PREPARATION  (R)\n")
cat(strrep("=", 70), "\n\n")

loan_applications <- read.csv(file.path(data_dir, "loan_applications.csv"), stringsAsFactors = FALSE)
credit_history    <- read.csv(file.path(data_dir, "credit_history.csv"),    stringsAsFactors = FALSE)
employment        <- read.csv(file.path(data_dir, "employment.csv"),        stringsAsFactors = FALSE)
payment_history   <- read.csv(file.path(data_dir, "payment_history.csv"),   stringsAsFactors = FALSE)

cat("--- Loaded Datasets ---\n")
cat(sprintf("  loan_applications: %6d rows x %d cols\n", nrow(loan_applications), ncol(loan_applications)))
cat(sprintf("  credit_history:    %6d rows x %d cols\n", nrow(credit_history),    ncol(credit_history)))
cat(sprintf("  employment:        %6d rows x %d cols\n", nrow(employment),        ncol(employment)))
cat(sprintf("  payment_history:   %6d rows x %d cols\n", nrow(payment_history),   ncol(payment_history)))

# ============================================================================
# 2. DATA CARD
# ============================================================================
print_data_card <- function(name, df) {
  cat(sprintf("\n%s\nDATA CARD: %s\n%s\n", strrep("-", 70), name, strrep("-", 70)))
  cat(sprintf("  Rows: %s    Columns: %d\n\n", format(nrow(df), big.mark = ","), ncol(df)))
  cat(sprintf("  %-30s %-12s %8s\n", "Column", "Type", "Missing"))
  cat(sprintf("  %s %s %s\n", strrep("-", 30), strrep("-", 12), strrep("-", 8)))
  for (col in names(df)) {
    cat(sprintf("  %-30s %-12s %8d\n", col, class(df[[col]])[1], sum(is.na(df[[col]]))))
  }
  cat("\n  Sample rows (first 3):\n")
  print(head(df, 3))
}

cat("\n", strrep("=", 70), "\n")
cat("DATA CARDS\n")
cat(strrep("=", 70), "\n")

print_data_card("loan_applications.csv", loan_applications)
print_data_card("credit_history.csv",    credit_history)
print_data_card("employment.csv",        employment)
print_data_card("payment_history.csv",   payment_history)

# ============================================================================
# 3. BASIC SUMMARY STATISTICS
# ============================================================================
cat("\n", strrep("=", 70), "\n")
cat("SUMMARY STATISTICS\n")
cat(strrep("=", 70), "\n")

num_cols <- function(df) df[, sapply(df, is.numeric), drop = FALSE]

cat("\n--- loan_applications (numeric) ---\n")
print(summary(num_cols(loan_applications)))

cat("\n--- loan_applications (categorical) ---\n")
print(table(loan_applications$loan_purpose))
print(table(loan_applications$property_type))

cat("\n--- credit_history (numeric) ---\n")
print(summary(num_cols(credit_history)))

cat("\n--- employment (numeric) ---\n")
print(summary(num_cols(employment)))

cat("\n--- employment (categorical) ---\n")
print(table(employment$employment_status))
print(table(employment$employer_type))

cat("\n--- payment_history (numeric) ---\n")
print(summary(num_cols(payment_history)))

cat("\n--- payment_history (categorical) ---\n")
print(table(payment_history$payment_status))

# ============================================================================
# 4. CLEAN & PARSE DATES
# ============================================================================
cat("\n", strrep("=", 70), "\n")
cat("DATA CLEANING\n")
cat(strrep("=", 70), "\n")

loan_applications$application_date <- as.Date(loan_applications$application_date)
payment_history$payment_date       <- as.Date(payment_history$payment_date)

valid_ids       <- unique(loan_applications$loan_id)
credit_history  <- credit_history[credit_history$loan_id %in% valid_ids, ]
employment      <- employment[employment$loan_id %in% valid_ids, ]
payment_history <- payment_history[payment_history$loan_id %in% valid_ids, ]

cat("  Date columns parsed, IDs validated.\n")

# ============================================================================
# 5. FEATURE ENGINEERING
# ============================================================================
cat("\n", strrep("=", 70), "\n")
cat("FEATURE ENGINEERING\n")
cat(strrep("=", 70), "\n")

# --- Payment behavior features ---------------------------------------------
pmt <- do.call(rbind, lapply(split(payment_history, payment_history$loan_id), function(d) {
  shortfall_ratio <- ifelse(d$amount_due > 0,
                            (d$amount_due - d$amount_paid) / d$amount_due, 0)
  data.frame(
    loan_id                  = d$loan_id[1],
    total_payments           = nrow(d),
    late_payment_count       = sum(d$days_late > 0, na.rm = TRUE),
    severe_delinquency_count = sum(d$days_late >= 60, na.rm = TRUE),
    avg_days_late            = mean(d$days_late, na.rm = TRUE),
    max_days_late            = max(d$days_late, na.rm = TRUE),
    total_shortfall          = max(sum(d$amount_due - d$amount_paid, na.rm = TRUE), 0),
    avg_shortfall_ratio      = mean(shortfall_ratio, na.rm = TRUE),
    on_time_count            = sum(d$payment_status == "On Time", na.rm = TRUE),
    stringsAsFactors = FALSE
  )
}))
rownames(pmt) <- NULL

pmt$late_payment_rate       <- ifelse(pmt$total_payments > 0,
                                       pmt$late_payment_count / pmt$total_payments, 0)
pmt$on_time_rate            <- ifelse(pmt$total_payments > 0,
                                       pmt$on_time_count / pmt$total_payments, 0)
pmt$severe_delinquency_flag <- as.integer(pmt$severe_delinquency_count > 0)

cat(sprintf("  Payment behavior features: %d\n", ncol(pmt) - 1))

# --- Credit features -------------------------------------------------------
credit <- credit_history

credit$fico_band <- ifelse(credit$credit_score >= 750, "Excellent",
                    ifelse(credit$credit_score >= 700, "Good",
                    ifelse(credit$credit_score >= 650, "Fair",
                    ifelse(credit$credit_score >= 600, "Poor", "Very Poor"))))

credit$utilization_cat <- ifelse(credit$credit_utilization <= 0.30, "Low",
                          ifelse(credit$credit_utilization <= 0.50, "Moderate",
                          ifelse(credit$credit_utilization <= 0.75, "High", "Very High")))

credit$has_bankruptcy  <- as.integer(credit$bankruptcies > 0)
credit$has_delinquency <- as.integer(credit$delinquencies_2yr > 0)
credit$high_inquiries  <- as.integer(credit$inquiries_6mo >= 3)

cat(sprintf("  Credit features:           %d\n", ncol(credit) - 1))

# --- Employment features ---------------------------------------------------
emp <- employment

emp$debt_service_ratio <- ifelse(emp$annual_income > 0,
                                  (emp$monthly_debt * 12) / emp$annual_income, 0)

emp$income_band <- ifelse(emp$annual_income >= 100000, "High",
                   ifelse(emp$annual_income >= 60000,  "Middle",
                   ifelse(emp$annual_income >= 35000,  "Low-Middle", "Low")))

emp$tenure_band <- ifelse(emp$years_employed >= 10, "Veteran",
                   ifelse(emp$years_employed >= 5,  "Established",
                   ifelse(emp$years_employed >= 2,  "Developing", "New")))

cat(sprintf("  Employment features:       %d\n", ncol(emp) - 1))

# ============================================================================
# 6. BUILD THE ANALYTICAL BASE TABLE
# ============================================================================
cat("\n", strrep("=", 70), "\n")
cat("BUILDING ANALYTICAL BASE TABLE\n")
cat(strrep("=", 70), "\n")

abt <- loan_applications
abt$application_date <- NULL

abt <- merge(abt, credit, by = "loan_id", all.x = TRUE)
abt <- merge(abt, emp,    by = "loan_id", all.x = TRUE)
abt <- merge(abt, pmt,    by = "loan_id", all.x = TRUE)

# Fill missing payment features
pmt_fill_cols <- setdiff(names(pmt), "loan_id")
for (col in pmt_fill_cols) {
  abt[[col]][is.na(abt[[col]])] <- 0
}
abt$debt_service_ratio[is.na(abt$debt_service_ratio)] <- 0

# One-hot encode categoricals
# loan_purpose
purposes <- unique(abt$loan_purpose[!is.na(abt$loan_purpose)])
for (p in purposes) {
  col_name <- paste0("purpose_", gsub("[^A-Za-z]", "", p))
  abt[[col_name]] <- as.integer(abt$loan_purpose == p)
}
abt$loan_purpose <- NULL

# property_type
props <- unique(abt$property_type[!is.na(abt$property_type)])
for (p in props) {
  col_name <- paste0("prop_", gsub("[^A-Za-z]", "", p))
  abt[[col_name]] <- as.integer(abt$property_type == p)
}
abt$property_type <- NULL

# employment_status
emp_statuses <- unique(abt$employment_status[!is.na(abt$employment_status)])
for (s in emp_statuses) {
  col_name <- paste0("emp_", gsub("[^A-Za-z]", "", s))
  abt[[col_name]] <- as.integer(abt$employment_status == s)
}
abt$employment_status <- NULL

# employer_type
emp_types <- unique(abt$employer_type[!is.na(abt$employer_type)])
for (t in emp_types) {
  col_name <- paste0("emplr_", gsub("[^A-Za-z]", "", t))
  abt[[col_name]] <- as.integer(abt$employer_type == t)
}
abt$employer_type <- NULL

# fico_band
fico_bands <- unique(abt$fico_band[!is.na(abt$fico_band)])
for (b in fico_bands) {
  col_name <- paste0("fico_", gsub("[^A-Za-z]", "", b))
  abt[[col_name]] <- as.integer(abt$fico_band == b)
}
abt$fico_band <- NULL

# utilization_cat
util_cats <- unique(abt$utilization_cat[!is.na(abt$utilization_cat)])
for (u in util_cats) {
  col_name <- paste0("util_", gsub("[^A-Za-z]", "", u))
  abt[[col_name]] <- as.integer(abt$utilization_cat == u)
}
abt$utilization_cat <- NULL

# income_band
inc_bands <- unique(abt$income_band[!is.na(abt$income_band)])
for (b in inc_bands) {
  col_name <- paste0("inc_", gsub("[^A-Za-z]", "", b))
  abt[[col_name]] <- as.integer(abt$income_band == b)
}
abt$income_band <- NULL

# tenure_band
ten_bands <- unique(abt$tenure_band[!is.na(abt$tenure_band)])
for (b in ten_bands) {
  col_name <- paste0("ten_", gsub("[^A-Za-z]", "", b))
  abt[[col_name]] <- as.integer(abt$tenure_band == b)
}
abt$tenure_band <- NULL

cat(sprintf("  Final ABT shape: %s rows x %d columns\n",
            format(nrow(abt), big.mark = ","), ncol(abt)))
cat(sprintf("  Features: %d  (excluding loan_id and defaulted)\n", ncol(abt) - 2))

# ============================================================================
# 7. DEFAULT DISTRIBUTION
# ============================================================================
cat("\n", strrep("=", 70), "\n")
cat("DEFAULT DISTRIBUTION IN ABT\n")
cat(strrep("=", 70), "\n")

default_tbl <- table(abt$defaulted)
cat(sprintf("  Current (0):   %5d  (%.1f%%)\n",
            default_tbl["0"], 100 * default_tbl["0"] / nrow(abt)))
cat(sprintf("  Defaulted (1): %5d  (%.1f%%)\n",
            default_tbl["1"], 100 * default_tbl["1"] / nrow(abt)))

# ============================================================================
# 8. SAVE
# ============================================================================
write.csv(abt, output_path, row.names = FALSE)
cat(sprintf("\nAnalytical Base Table saved to: %s\n", output_path))
cat(strrep("=", 70), "\n")
cat("STEP 2 COMPLETE\n")
cat(strrep("=", 70), "\n")
